### ISSUE TITLE
Stale InMemoryCache.optimisticData selection key/value cache when using optimisticResponse

### ISSUE THREAD

**Issue by markphilpot** [apollographql/apollo-client#7894]
I've verified this behavior with 3.2.x, 3.3.x and 3.4.0-beta15 using a react 16.10.x tool chain.

**tl;dr** It is possible to have old stale data remain in the optimisticData's selection Key cache such that every time an `optimisticResponse` mutation is provided, the UX will flash the stale state.

**Terminology**

- inMemoryCache
	- this.data: EntityStore
	- this.optimisticData: EntityStore

- storeReader.js
	- this.executeSelectionSet
	- this.knownResults

From what I can tell `executeSelctionSet` operates against an intermediate field based cache `knownResults` (or something similar) that has field lookup values.  It doesn't appear that this cache is ever invalidated when `client.writeQuery` is used.

**Scenario**

ComponentA backed by graphql object with scalar field `title`
- modifying `title` uses an `optimisticResponse` using `useMutation` hook

ComponentB backed by ROOT_QUERY field `containerContents` with a field `cards` which is a list of object references
- Adding object to list `cards` uses a direct `client.writeQuery`

**Timeline**

1. ComponentA `title` = "Foo", ComponentB is showing an empty list
2. Update `title`
3. React Render Pass when `optimisticData` is layered in
	1. The field selection cache of `optimisticData` now has an empty list state
4. Add object to `cards` using `client.writeQuery`
	1. The correct `cards` list (length 1) is represented in both `data` and `optimisticData` (verified by inspection)
	2. *Important to note* -- The field selection cache of `optimisticData` still has `cards` as empty list
5. Update `title` again
	1. UI flashes old empty `cards` state when `optimisticData` is layered in (verified that when `this.executeSelectionSet` is called against the `optimisticData` store, it returns an empty list despite the underlying data matching.

**Intended outcome:**

Using the optimistic UI pattern doesn't flash old/stale state when previous values were "captured" by the field cache.

**Actual outcome:**

Using the optimistic UI pattern flashes old/stale data if a react render pass occurred when the `optimisticData` layer was used

**Versions**

  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 15.11.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.6.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 89.0.4389.90
    Firefox: 86.0.1
    Safari: 14.0.3

**Comment by benjamn** [comment]
@markphilpot Thanks for digging into this!

I'm cautiously optimistic that #7887 (slated for the next minor release, v3.4) may improve this behavior. Can you try updating to `@apollo/client@3.4.0-beta.16` (just published)?

If that doesn't solve the problem, it would be very helpful if could put your scenario into a runnable [reproduction](https://github.com/apollographql/react-apollo-error-template/).

**Comment by markphilpot** [comment]
Testing this morning seems to indicate `@apollo/client@3.4.0-beta.16` does indeed fix the issue!  Excellent work! 🙏 

Closing issue.
